### DISCUSSION TITLE
Add MuJoCo to ROS Buildfarm

### DISCUSSION THREAD

**Discussion by peterdavidfagan** [google-deepmind/mujoco#1387] [General]
I am starting to develop MuJoCo simulation environments that interface with ROS 2. I have a working example in Python but need to switch to the C API for performance reasons. Currently MuJoCo doesn't appear to be released under the ROS 2 package management ecosystem, I suggest that MuJoCo is added as a package so it can be used more readily by the ROS community. 

I guess this is not strictly necessary but it makes handling the MuJoCo dependency a lot cleaner when developing ROS packages. Looks like it may take not insignificant work to add this.

**Comment by traversaro** [comment]
Note that mujoco is available in recent version of debian, even if the version is relatively old: https://packages.debian.org/source/sid/mujoco, so perhaps if mujoco is added to the buildfarm it could make sense to add it as vendor package.

**Reply by peterdavidfagan** [reply]
Nice, thanks @traversaro, I wasn't aware of this debian. I have been manually downloading the release to my machine and configuring CMake to point towards shared library files (still in dev as I only just made the switch), agree on it being added as a vendor package.

For now I am going to continue with what I have but I hope in future I'll be able to package my software with MuJoCo in a way that aligns with the ROS ecosystem as I am creating a ROS package.

**Comment by peterdavidfagan** [comment]
As an update to this thread, for now I add the latest release files to my repository and link them to the project. Not ideal but it gets the job done for my use case: https://github.com/peterdavidfagan/mujoco_ros.

**Reply by peterdavidfagan** [reply]
Closing for now as build farm integration isn't fully necessary and I don't have the capacity to work on this right now. Would be happy to continue the discussion again at some point if other ROS users think this would be useful.

**Reply by DavidPL1** [reply]
You really should not add the latest release files to your repo. Why don't you reproduce what is done to build Simulate [here](https://github.com/google-deepmind/mujoco/blob/03a8fa9ca90a8bed1c5bca34227465bde98e07f6/simulate/cmake/SimulateDependencies.cmake#L40C1-L59C2) and/or use a MUJOCO_DIR environment variable to find an unpacked tar of a compatible release in your CMakeLists.txt?

**Reply by peterdavidfagan** [reply]
I agree I don't think what I have done is best practice, I'll add an issue to migrate to the findorfetch utility.

**Reply by githubs-user** [reply]
What is the problem with using the latest release file in the repo? 
And how we can add the compatible release of MuJoCo to our ROS2 package?

**Reply by peterdavidfagan** [reply]
> What is the problem with using the latest release file in the repo?

Generally you want to handle dependencies with CMake, manually adding a folder for release files is bad practice as it adds unnecessary bloat to the package, imagine having to do this for multiple dependencies.

> And how we can add the compatible release of MuJoCo to our ROS2 package?

The link shared by @DavidPL1 demonstrates how to findorfetch MuJoCo, you can add this to your CMakeLists.txt file and appropriately link against MuJoCo to add this dependency.

**Reply by peterdavidfagan** [reply]
Hi @DavidPL1, 

I went with your suggestion and updated my CMakeLists.txt to use FindOrFetch. Thanks for the advice on this.
